osidb/models.py

```diff
@@ -696,6 +732,16 @@ def _validate_cvss3(self):
                 "CVSSv3 score is missing",
             )
 
+    def _validate_major_incident_state(self):
+        """
+        Checks that a flaw has a valid Major Incident state.
+        """
+        if self.major_incident_state == self.FlawMajorIncident.MI_INVALID:
+            self.alert(
+                "mi_state_invalid",
+                "A flaw does not have a valid Major Incident state.",
+            )
```

Comment:
I think having an `INVALID` value is a mistake:

* This means clients can set `major_incident_state` to `INVALID` by hand through the OSIDB API, this should not happen and all values should be valid
* We could set this to be a hard error, i.e. we raise a `ValidationError` which will be non-blocking at ingestion but this sends mixed signals to clients, the schema tells them `INVALID` is a valid value to send in an API call but the server will refuse to accept it via REST API.

Instead, we could simply have invalid states be the default (`UNKNOWN`), clients will complain when their `major_incident_state` is not correct either way, regardless of it being `UNKNOWN` or `INVALID`.

Reply:
Good points. Client should definitely not be able to set an invalid state. Therefore if we go with it there should be a blocking validation checking that this value is not present (could be set by `bzimport` only).

With @jobselko we were discussing how to approach the situation when there is a combination of the Bugzilla flag values (it is two independent flags) which does not make sense/is not clear what is meant by it. We came with two possible solutions.

- I was first suggesting to give priority to some values (generally to the regular Major Incident above the CISA one) and simply map all the combinations to the valid values only. The downside is that when synced back we would basically fix the combination possibly not in the way the user intended to.
- Therefore Jitka came with the solution of the `INVALID` value which explicitly means that the original Bugzilla data must be fixed (either in Bugzilla or by setting a valid value in OSIDB) before we can do anything with the flaw.

True that this is really Bugzilla related thing and should be temporary (and probably very rare) and it would appear in the schema as a possible values (though not settable).

Your suggestion to use `UNKNOWN` instead is interesting but I see an issue there. The `UNKNOWN` means that there are no Major Incident flags present in Bugzilla (not sure if the name is great then). Mapping the invalid combination to it would result in erasing the flags on flaw update which would be unexpected from the user point of view and possibly harmful - even though it was set incorrectly the Major Incident is very serious matter so we should rather force the user to fix ti instead of erasing it.

Reply:
We could have `bbsync` set up so that if the state is `UNKNOWN` we simply write back whatever we have in `meta_attr`.

But this is a suggestion, having a `ValidationError` is acceptable.

Reply:
I would use `ValidationError` because we should differentiate between `UNKNOWN` and `INVALID` (if something is wrong in BZ, we just cannot rewrite it by a different value). `UNKNOWN` could be rewritten as `NOT_SET` to better describe the current state. I will also change `is_major_incident` field to property.

All functionality should be independent on `FlawMeta`, and I will not be touching `FlawMeta` for now.

@Elkasitu Do you agree with this proposal?

Reply:
Renaming `UNKNOWN` to `NOT_SET` makes a lot of sense. It actually means that no flag was raised so the status is known. Then I think it becomes self-explanatory that no other status should be mapped to it.

Reply:
I believe `blank=True` might be a better choice then? I guess that it depends on whether we consider a `Flaw` complete without a `major_incident_state` or not

Reply:
In fact we have a lot of cases already in which the field is `blank=True` and we assign the empty value to an enum value like so: `NOVALUE = ""`, `FlawImpact` is one such case

Reply:
@Elkasitu All changes except the `is_major_incident` change to the property are done. I tried to make it a property, but many tests failed, and fixing them would take me some time. So can I leave it as it is now and make this change later?

Reply:
Sure

Reply:
Thank you. I will create a Jira for it.